### DISCUSSION TITLE
Generative bot response in combination with rail in conversation flow

### DISCUSSION THREAD

**Discussion by amoryblaine1** [NVIDIA-NeMo/Guardrails#111] [General]
Hello! Me and two other developers at my company are actively experimenting with NeMo-Guardrails in an attempt to improve our current agent architecture. We've been working with Guardrails daily for the past 2 weeks and keep running into this issue and can't seem to figure out a solution in the documentation....

We simply want to combine a generative response with our rails within the conversation flow.

For example:

def user express feelings
     "Today is terrible. My dog died!"
     "I had a horrible day.... I was struck by lightning!"

def bot express agreement
    "Remember, the day can only get better from here!"

define flow encourage user
    user express feelings
    bot express agreement

What I want to do is.... generate a custom "empathy" response based on "user express feelings", then have the flow continue and trigger our rail "bot express agreement"

So that the final response would be something along the lines of:

BOT: "I'm so sorry to hear about your dog. Remember, the day can only get better from here!"

If anyone has any ideas, I would be very grateful to hear them!

**Comment by trebedea** [comment]
Hi,

So you want to do something like this?

```
def user express feelings
    "Today is terrible. My dog died!"
    "I had a horrible day.... I was struck by lightning!"

def user does not express feelings
    "Hi there! How are you?"
    "I need some advice with planning a holiday"
    "Are you good at programming?"
    "Can you tell me a joke?"

def bot express agreement
    "Remember, the day can only get better from here!"

define flow encourage user
    user express feelings
    bot express empathy towards user feeling
    bot express agreement
```

I've just added a new canonical form, in order for the bot not to be biased towards over-generating `user express feelings` .
Otherwise, you can add multiple bot consecutive bot canonical forms and for the ones that do not have a specified output (e.g. `bot express empathy towards user feeling`), the LLM is called to sample a response given the whole conversation as a context, railed by this specific canonical form as the last line in the context. 

You can also branch the flow to provide different continuations for positive or negative emotions, but you need to define different user canonical forms and branch on those.

**Reply by amoryblaine1** [reply]
Thank you for your response! Really appreciate it. This makes sense and is indeed a potentially workable solution. Much appreciated.
